### PR TITLE
ONNX : ReflecitonPad2d and ZeroPad2d Subgraph fusion added

### LINKED ISSUE #16971: Cannot load the onnx model converted from pytorch
##### System information (version)

- OpenCV => OpenCV => 4.3.0(e6f32c6a69043456a806a4e802ee3ce7b7059c93)
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017 64 bits

##### Detailed description

The decoder model converted from [this project](https://github.com/nianticlabs/monodepth2) do not work.

##### Steps to reproduce

1. install pytorch 1.4.0(cpu version)
2. clone the project
3. Navigate to the folder of the project, create a file called to_onnx.py

```
from __future__ import absolute_import, division, print_function

import os
import sys
import glob
import numpy as np
import PIL.Image as pil
import matplotlib as mpl
import matplotlib.cm as cm

import torch
from torchvision import transforms, datasets

import networks
from layers import disp_to_depth
from utils import download_model_if_doesnt_exist

device = torch.device("cpu")

model_name = "mono+stereo_640x192"
model_path = os.path.join("models", "mono+stereo_640x192")
print("-> Loading model from ", model_path)
encoder_path = os.path.join(model_path, "encoder.pth")
depth_decoder_path = os.path.join(model_path, "depth.pth")

# LOADING PRETRAINED MODEL
print("   Loading pretrained encoder")
encoder = networks.ResnetEncoder(18, False)
loaded_dict_enc = torch.load(encoder_path, map_location=device)

# extract the height and width of image that this model was trained with
feed_height = loaded_dict_enc['height']
feed_width = loaded_dict_enc['width']

filtered_dict_enc = {k: v for k, v in loaded_dict_enc.items() if k in encoder.state_dict()}
encoder.load_state_dict(filtered_dict_enc)
encoder.to(device)
encoder.eval()

import onnx
import onnxruntime

def test_model_accuracy(export_model_name, raw_output, input):    
    ort_session = onnxruntime.InferenceSession(export_model_name)

    def to_numpy(tensor):
        return tensor.detach().cpu().numpy() if tensor.requires_grad else tensor.cpu().numpy()

    # compute ONNX Runtime output prediction
    ort_inputs = {ort_session.get_inputs()[0].name: to_numpy(input)}
    ort_outs = ort_session.run(None, ort_inputs)	

    # compare ONNX Runtime and PyTorch results
    np.testing.assert_allclose(to_numpy(raw_output), ort_outs[0], rtol=1e-03, atol=1e-05)

    print("Exported model has been tested with ONNXRuntime, and the result looks good!")		

def export_model(model, input, export_model_name):
    torch.onnx.export(model, input, export_model_name, verbose=False, export_params=True, opset_version=11)	
    onnx_model = onnx.load(export_model_name)    
    onnx.checker.check_model(onnx_model)
    graph_output = onnx.helper.printable_graph(onnx_model.graph)
    with open("graph_output.txt", mode="w") as fout:
        fout.write(graph_output)

export_model_name = "encoder_" + model_name + '.onnx'
input_image = pil.open("assets/runner.jpg").convert('RGB')
input_image = input_image.resize((feed_width, feed_height), pil.LANCZOS)
input_image = transforms.ToTensor()(input_image).unsqueeze(0)

features = encoder(input_image)
export_model(encoder, input_image, export_model_name)
test_model_accuracy(export_model_name, features, input_image)

#print("   Loading pretrained decoder")
depth_decoder = networks.DepthDecoder(
    num_ch_enc = encoder.num_ch_enc, scales=range(4))

loaded_dict = torch.load(depth_decoder_path, map_location=device)
depth_decoder.load_state_dict(loaded_dict)

depth_decoder.to(device)
depth_decoder.eval()

decoder_output = depth_decoder(features)

export_model_name = "decoder_" + model_name + '.onnx'

export_model(depth_decoder, features, export_model_name)
test_model_accuracy(export_model_name, decoder_output, features)
```
4. Change the class ResnetEncoder in networks/resnet_encoder.py to

```
class ResnetEncoder(nn.Module):
    """Pytorch module for a resnet encoder
    """
    def __init__(self, num_layers, pretrained, num_input_images=1):
        super(ResnetEncoder, self).__init__()

        self.num_ch_enc = np.array([64, 64, 128, 256, 512])

        resnets = models.resnet18
        self.encoder = resnets(pretrained)	        

    def forward(self, input_image):         
        x = (input_image - 0.45) / 0.225
        x = self.encoder.conv1(x)
        x = self.encoder.bn1(x)		        
        output1 = self.encoder.relu(x)
        output2 = self.encoder.layer1(self.encoder.maxpool(output1))
        output3 = self.encoder.layer2(output2)
        output4 = self.encoder.layer3(output3)
        output5 = self.encoder.layer4(output4)        	
		
        return torch.cat((output1.flatten(), output2.flatten(), output3.flatten(), output4.flatten(), output5.flatten()))        
```

5. Change the class DepthDecoder in networks/depth_decoder.py to

```
class DepthDecoder(nn.Module):
    def __init__(self, num_ch_enc, scales=range(4), num_output_channels=1, use_skips=True):
        super(DepthDecoder, self).__init__()

        self.num_output_channels = num_output_channels
        self.use_skips = use_skips
        self.upsample_mode = 'nearest'
        self.scales = scales

        self.num_ch_enc = num_ch_enc
        self.num_ch_dec = np.array([16, 32, 64, 128, 256])

        # decoder
        self.convs = OrderedDict()        
        for i in range(4, -1, -1):
            # upconv_0
            num_ch_in = self.num_ch_enc[-1] if i == 4 else self.num_ch_dec[i + 1]
            num_ch_out = self.num_ch_dec[i]
            self.convs[("upconv", i, 0)] = ConvBlock(num_ch_in, num_ch_out)

            # upconv_1
            num_ch_in = self.num_ch_dec[i]
            if self.use_skips and i > 0:
                num_ch_in += self.num_ch_enc[i - 1]
            num_ch_out = self.num_ch_dec[i]
            self.convs[("upconv", i, 1)] = ConvBlock(num_ch_in, num_ch_out)

        for s in self.scales:
            self.convs[("dispconv", s)] = Conv3x3(self.num_ch_dec[s], self.num_output_channels)

        self.decoder = nn.ModuleList(list(self.convs.values()))
        self.sigmoid = nn.Sigmoid()	        

    def forward(self, input_features):      
        #simplify the forward functions a lot, only post the codes which cause the issue  
        #the output processed by self.convs[("upconv", 4, 0)] will caused the error message
        return self.convs[("upconv", 4, 0)](input_features[2826240:2887680].reshape(1, 512, 6, 20))
```
6. Load the model by opencv

```
std::string const model_folder("C:/Users/yyyy/programs/Qt/app_to_simplify_ml/"
                                   "cv_projects/clients/mobile_depth_estimation/monodepth2");
cv::dnn::Net net = cv::dnn::readNet(model_folder + "/decoder_mono+stereo_640x192.onnx");
```

Both of the decoder and encoder models passed the test.

To save your troubles, you can have the graph,model and codes(to_onnx.py, resnet_encoder.py, depth_decoder.py) from [google drive](https://drive.google.com/open?id=1XNgdQhdQ8WW2R_ApyuTtJ8ituPhsGED2).

##### Issue submission checklist

Cannot load the encoder model after convert,  got error messages

**OpenCV: terminate handler is called! The last OpenCV error is:OpenCV(4.3.0-pre) Error: Assertion failed (shapeIt != outShapes.end()) in cv::dnn::dnn4_v20200310::ONNXImporter::populateNet, file C:\Users\yyyy\programs\Qt\3rdLibs\opencv\dev\opencv\modules\dnn\src\onnx\onnx_importer.cpp, line 1116**

### PR DESCRIPTION
**Pull request by ashishkrshrivastava**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/740

resolves #16971 

```
opencv_extra=padfusion
```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### This PR adds
SubgraphFusion for 

- [x] ReflictionPad2d (pytorch)
- [x] ZeroPad2d (pytorch)

Screenshot from 2020-04-05 11-29-11

Screenshot from 2020-04-05 11-29-31

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2020.1.0:16.04
build_image:Custom Win=openvino-2020.1.0
build_image:Custom Mac=openvino-2020.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 1]
```diff
@@ -290,9 +290,112 @@ class ResizeSubgraph2 : public ExtractScalesSubgraph
...
+        int constantEnds = addNodeToMatch("Constant");
+        int constantSteps = addNodeToMatch("Constant");
+
+        int sliceInputs[] = {reshapeConcat, constantStart, constantEnds, constantAxes, constantSteps};
+
+        int slice = addNodeToMatch("Slice", std::vector<int>(&sliceInputs[0], &sliceInputs[0] + 5));
+        int transpose = addNodeToMatch("Transpose", slice);
+        int constantRT = addNodeToMatch("Constant");
+        int reshapeT = addNodeToMatch("Reshape", transpose, constantRT);
+        castRT = addNodeToMatch("Cast", reshapeT);
```
Comment (dkurt) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:323]: Can onnx_importer computer this tensor?

Comment (ashishkrshrivastava) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:323]: @dkurt No, it cannot. 

I tried some experiments, made some changes but ended up with this error at slice layer. 

```
cv2.error: OpenCV(3.4.10-dev) /home/ashish/opencv/modules/dnn/src/onnx/onnx_importer.cpp:485: error: (-2:Unspecified error) in function 'void cv::dnn::experimental_dnn_34_v17::ONNXImporter::populateNet(cv::dnn::experimental_dnn_34_v17::Net)'
> Slice layer only supports steps = 1 (expected: 'countNonZero(step_blob != 1) == 0'), where
>     'countNonZero(step_blob != 1)' is 1
> must be equal to
>     '0' is 0
```
Because slice layer has inputs shown in the image.
Screenshot from 2020-04-06 10-56-45

**Change I made**
Change 1. allow shape to take inpshape from constBlob
```cpp
else if (layer_type == "Shape")
        {
            CV_Assert(node_proto.input_size() == 1);
            shapeIt = outShapes.find(node_proto.input(0));

            MatShape inpShape = shapeIt != outShapes.end() ? shapeIt->second : shape(getBlob(node_proto, constBlobs, 0));
            Mat shapeMat(inpShape.size(), 1, CV_32S);
            for (int j = 0; j < inpShape.size(); ++j)
                shapeMat.at<int>(j) = inpShape[j];
            shapeMat.dims = 1;

            constBlobs.insert(std::make_pair(layerParams.name, shapeMat));
            continue;
        }
```
from 
```cpp
else if (layer_type == "Shape")
        {
            CV_Assert(node_proto.input_size() == 1);
            shapeIt = outShapes.find(node_proto.input(0));
            CV_Assert(shapeIt != outShapes.end());
            MatShape inpShape = shapeIt->second;

            Mat shapeMat(inpShape.size(), 1, CV_32S);
            for (int j = 0; j < inpShape.size(); ++j)
                shapeMat.at<int>(j) = inpShape[j];
            shapeMat.dims = 1;

            constBlobs.insert(std::make_pair(layerParams.name, shapeMat));
            continue;
        }
```


Change 2. make all input datatypes same. 
input 1 from cast was int and input 2 from constantofshape was float32, i guess, so was giving error. 
```cpp
else if (layer_type == "Concat")
        {
            bool hasVariableInps = false;
            for (int i = 0; i < node_proto.input_size(); ++i)
            {
                if (layer_id.find(node_proto.input(i)) != layer_id.end())
                {
                    hasVariableInps = true;
                    break;
                }
            }

            if (!hasVariableInps)
            {
                std::vector<Mat> inputs(node_proto.input_size()), concatenated;
                for (size_t i = 0; i < inputs.size(); ++i)
                {
                    inputs[i] = getBlob(node_proto, constBlobs, i);
                    inputs[i].convertTo(inputs[i], CV_32F);
                }
                runLayer(layerParams, inputs, concatenated);

                CV_Assert(concatenated.size() == 1);
                constBlobs.insert(std::make_pair(layerParams.name, concatenated[0]));
                continue;
            }
        }
```
from 
```cpp
else if (layer_type == "Concat")
        {
            bool hasVariableInps = false;
            for (int i = 0; i < node_proto.input_size(); ++i)
            {
                if (layer_id.find(node_proto.input(i)) != layer_id.end())
                {
                    hasVariableInps = true;
                    break;
                }
            }

            if (!hasVariableInps)
            {
                std::vector<Mat> inputs(node_proto.input_size()), concatenated;
                for (size_t i = 0; i < inputs.size(); ++i)
                {
                    inputs[i] = getBlob(node_proto, constBlobs, i);
                }
                runLayer(layerParams, inputs, concatenated);

                CV_Assert(concatenated.size() == 1);
                constBlobs.insert(std::make_pair(layerParams.name, concatenated[0]));
                continue;
            }
        }
```


change 3. if data is less than min or greater than max, dont give error. 

```cpp
void convertInt64ToInt32(const T1& src, T2& dst, int size)
{
    for (int i = 0; i < size; i++)
    {
        if (src[i] < std::numeric_limits<int32_t>::min())
        {
            dst[i] = saturate_cast<int32_t>(std::numeric_limits<int32_t>::min());
            continue;
        }
        if (src[i] > std::numeric_limits<int32_t>::max())
        {
            dst[i] = saturate_cast<int32_t>(std::numeric_limits<int32_t>::max());
            continue;
        }
        dst[i] = saturate_cast<int32_t>(src[i]);
    }
}
```
from 
```cpp
void convertInt64ToInt32(const T1& src, T2& dst, int size)
{
    for (int i = 0; i < size; i++) {
        if (src[i] < std::numeric_limits<int32_t>::min() || src[i] > std::numeric_limits<int32_t>::max()) {
            CV_Error(Error::StsOutOfRange, "Input is out of OpenCV 32S range");
        }
        dst[i] = saturate_cast<int32_t>(src[i]);
    }
}
```

And after all that slice node gave above mentioned error.

**Comment by dkurt** [comment]
@ashishkrshrivastava, May I ask you to try this patch: https://github.com/dkurt/opencv/commit/0f92db7afc6382c4b8cd684e096557bc4050ae54? It solves some imported problems and enables ZeroPad2d by constants folding rather than graph fusion. Can you test `ReflectionPad2d` with that as well and extend if needed? Thanks!

**Comment by ashishkrshrivastava** [comment]
> Can you test ReflectionPad2d with that as well and extend if needed?

@dkurt I have added changes with some modification for ReflectionPad2d

File: modules/dnn/test/test_onnx_importer.cpp [hunk 2]
```diff
@@ -490,6 +490,16 @@ TEST_P(Test_ONNX_layers, LSTM_bidirectional)
     testONNXModels("lstm_bidirectional", npy, 0, 0, false, false);
 }
 
+TEST_P(Test_ONNX_layers, Pad2d_Unfused)
+{
+    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NN_BUILDER_2019)
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NN_BUILDER);
+    if (backend == DNN_BACKEND_INFERENCE_ENGINE_NGRAPH)
+        applyTestTag(CV_TEST_TAG_DNN_SKIP_IE_NGRAPH);
```
Comment (dkurt) [inline modules/dnn/test/test_onnx_importer.cpp:498]: Have you tested it with this backends? If not, let's try to remove and see the results on buildbot

Comment (ashishkrshrivastava) [reply modules/dnn/test/test_onnx_importer.cpp]: Done!
